### PR TITLE
bpo-1875, bpo-32477: Raise SyntaxError in invalid blocks that will be optimized away.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Also simplify the code generator and peepholer for "if" and "while" with
constant condition.
Also optimize more cases of conditional jumps with constant condition.

https://bugs.python.org/issue1875

https://bugs.python.org/issue32477

### CODE REVIEW

File: Python/compile.c [hunk 1]
```diff
@@ -2539,36 +2550,23 @@ static int
...
+        if (next == NULL)
             return 0;
-        VISIT_SEQ(c, stmt, s->v.If.body);
-        if (asdl_seq_LEN(s->v.If.orelse)) {
-            ADDOP_JREL(c, JUMP_FORWARD, end);
-            compiler_use_next_block(c, next);
-            VISIT_SEQ(c, stmt, s->v.If.orelse);
-        }
+    }
+    if (!compiler_jump_if(c, s->v.If.test, next, 0))
```
Comment (pablogsal) [inline Python/compile.c]: Nit: could you add braces? (PEP7)

File: Lib/test/test_syntax.py [hunk 2]
```diff
@@ -698,17 +698,35 @@ def test_break_outside_loop(self):
 
     def test_yield_outside_function(self):
         self._check_error("if 0: yield",               "outside function")
-        self._check_error("class C:\n    if 0: yield", "outside function")
+        self._check_error("if 1: pass\nelse: yield",   "outside function")
```
Comment (pablogsal) [inline Lib/test/test_syntax.py]: Could you add a test for the pattern in https://bugs.python.org/issue37289 ir you prefer to do it in another PR?

File: Python/compile.c [hunk 3]
```diff
@@ -2658,47 +2656,32 @@ compiler_async_for(struct compiler *c, stmt_ty s)
...
     loop = compiler_new_block(c);
     end = compiler_new_block(c);
-    if (constant == -1) {
-        anchor = compiler_new_block(c);
-        if (anchor == NULL)
-            return 0;
-    }
-    if (loop == NULL || end == NULL)
+    anchor = compiler_new_block(c);
+    if (loop == NULL || end == NULL || anchor == NULL)
```
Comment (pablogsal) [inline Python/compile.c:2772]: Nit: PEP7

File: Python/compile.c [hunk 4]
```diff
@@ -2658,47 +2656,32 @@ compiler_async_for(struct compiler *c, stmt_ty s)
...
-        orelse = NULL;
 
     compiler_use_next_block(c, loop);
     if (!compiler_push_fblock(c, WHILE_LOOP, loop, end))
         return 0;
-    if (constant == -1) {
-        if (!compiler_jump_if(c, s->v.While.test, anchor, 0))
-            return 0;
-    }
+    if (!compiler_jump_if(c, s->v.While.test, anchor, 0))
```
Comment (pablogsal) [inline Python/compile.c:2783]: Could you add braces?

File: Python/compile.c [hunk 5]
```diff
@@ -2539,36 +2550,23 @@ static int
...
-            next = compiler_new_block(c);
-            if (next == NULL)
-                return 0;
-        }
-        else
-            next = end;
-        if (!compiler_jump_if(c, s->v.If.test, next, 0))
+    if (asdl_seq_LEN(s->v.If.orelse)) {
+        next = compiler_new_block(c);
+        if (next == NULL)
```
Comment (pablogsal) [inline Python/compile.c:2667]: PEP7

File: Misc/NEWS.d/next/Core and Builtins/2019-06-15-19-01-05.bpo-1875.IfLQg1.rst [hunk 6]
```diff
@@ -0,0 +1,3 @@
+A :exc:`SyntaxError` is now always raised if a code blocks that will be
```
Comment (pablogsal) [inline Misc/NEWS.d/next/Core and Builtins/2019-06-15-19-01-05.bpo-1875.IfLQg1.rst:1]: 'if code blocks' or 'if a code block'

**Review by pablogsal — APPROVED** [review]
Thanks Serhiy!

I left some minor comments

**Comment by pablogsal** [comment]
It seems that this is causing some errors with trace functions

File: Lib/test/test_syntax.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Lib/test/test_syntax.py]: I think that test is unrelated.

**Review by pablogsal — CHANGES_REQUESTED** [review]
You need to update this test:

https://github.com/python/cpython/blob/master/Lib/test/test_sys_settrace.py#L62

File: Python/compile.c (hunk 5 above)
Comment (serhiy-storchaka) [reply Python/compile.c:2667]: It is not a new code, I just removed some code and reduced an indentation. `return 0;` without braces is uses every where in this file, so I think that breaking uniformity in the file can add more harm than disobeying this particular part of PEP 7.

Comment (pablogsal) [reply Python/compile.c:2667]: I agree, disregard the PEP7 comments then.

File: Python/compile.c [hunk 7]
```diff
@@ -2448,10 +2447,22 @@ compiler_jump_if(struct compiler *c, expr_ty e, basicblock *next, int cond)
         /* fallback to general implementation */
         break;
     }
-    default:
+    default: {
+        if (e->kind == Constant_kind) {
```
Comment (pablogsal) [inline Python/compile.c:2558]: Shouldn't this be it's on case in the previous switch?

(case Constant_kind: )

Comment (serhiy-storchaka) [reply Python/compile.c:2558]: Good point!

**Comment by pablogsal** [comment]
@serhiy-storchaka I think we should merge this sooner than later so it can be in the next 3.8 beta release and people can interact with it more. If you don't have time I can push to the PR to fix the failing test (https://github.com/python/cpython/blob/master/Lib/test/test_sys_settrace.py#L62) if you prefer :)

**Comment by serhiy-storchaka** [comment]
No, this PR is not ready yet.

File: Python/compile.c [hunk 8]
```diff
@@ -2646,48 +2657,24 @@ static int
...
+    next = end = compiler_new_block(c);
     if (end == NULL)
         return 0;
 
-    constant = expr_constant(s->v.If.test);
-    /* constant = 0: "if 0"
-     * constant = 1: "if 1", "if 2", ...
-     * constant = -1: rest */
-    if (constant == 0) {
-        BEGIN_DO_NOT_EMIT_BYTECODE
```
Comment (pablogsal) [inline Python/compile.c:2660]: You can remove this macro and the attribute in the compiler struct if is not used anymore

**Comment by markshannon** [comment]
This change was implemented as part of PEP 626, so this PR is now obsolete.
